### PR TITLE
Optimize formalTypes

### PR DESCRIPTION
**Pull request by retronym**

### CODE REVIEW

**Comment by diesalbla** [comment]
It may be worth it to add into the `List` companion object something like the [`replicate`](https://hackage.haskell.org/package/base-4.12.0.0/docs/Prelude.html#v:replicate) method, 
```Scala
def replicate[A](num: Int, elem: a): List[A]
```

**Review by diesalbla — APPROVED** [review]

**Comment by diesalbla** [comment]
Can we re-target to 2.12?

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 1]
```diff
@@ -50,11 +50,23 @@ trait Infer extends Checkable {
...
+      val result = mutable.ListBuffer[Type]()
+      var fs = formals
+      while ((fs ne Nil) && (fs.tail ne Nil)) {
+        result.addOne(fs.head)
+        fs = fs.tail
+      }
+      var i = 0
+      val n = numArgs - numFormals + 1
+      while (i < n) {
+        result += lastType
```
Comment (hrhino) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: `lastType` doesn't change here but it's also a `def` (and calls `List#last`)

**Comment by hrhino** [comment]
@diesalbla adding what to `List`?

**Comment by diesalbla** [comment]
@hrhino A method to build a list of length N, where all the positions are the same value.

**Comment by hrhino** [comment]
Ah. Yeah. We have `fill`, but its parameter is by name, so not quite what we need.

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala (hunk 1 above)
Comment (retronym) [reply src/compiler/scala/tools/nsc/typechecker/Infer.scala]: I've moved it to a `val` inside `if (expandLast)`

**Comment by retronym** [comment]
I've also added a utility method `fillList` and used that. Note that `ListBuffer.prependToList` is efficient, it shares the input list.

**Comment by retronym** [comment]
@diesalbla I'm going to be targetting more changes primary to 2.13, and doing backports selectively if any prove to have a compelling performance benefit. The merges from 2.12.x to 2.13.x are already pretty tricky.

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 2]
```diff
@@ -49,12 +49,21 @@ trait Infer extends Checkable {
...
-      formals1.init ::: expanded(numArgs - numFormals + 1)
-    else
+    if (expandLast) {
+      // extract the T from T*
+      val lastType = formals1.last.dealiasWiden.typeArgs.head
+
+      val n = numArgs - numFormals + 1
+      // Optimized version of: formals1.init ::: List.fill(n)(lastType)
+      val result = mutable.ListBuffer[Type]()
+      var fs = formals
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: was: `formals1` (for the test error)

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Infer.scala]: Why we need `x'` syntax to scope these init value contexts. Elsewhere there was a shadowed x that needed either warn-unused or lint value-param-shadow.

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 3]
```diff
@@ -49,12 +49,21 @@ trait Infer extends Checkable {
...
+      val lastType = formals1.last.dealiasWiden.typeArgs.head
+
+      val n = numArgs - numFormals + 1
+      // Optimized version of: formals1.init ::: List.fill(n)(lastType)
+      val result = mutable.ListBuffer[Type]()
+      var fs = formals1
+      while ((fs ne Nil) && (fs.tail ne Nil)) {
+        result.addOne(fs.head)
+        fs = fs.tail
+      }
```
Comment (hrhino) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala:64]: If you want to be very micro about your optimizations, after this loop, `fs.head.dealiasWiden.typeArgs.head` is your `lastType`. Also `fs` should never be `Nil`.

**Review by hrhino — APPROVED** [review]

**Comment by SethTisue** [comment]
@retronym ready for merge?

**Pull request merged**